### PR TITLE
Overhaul 1 - Initial Multifunction support, some compatibility with SciML

### PR DESCRIPTION
**Pull request by VarLad**
This PR aims for multifunction encoding as well as introduces the DQCType.
The work in this PR is incomplete as compatibility with types from SciMLBase is blocked by Zygote not being able to handle array mutations. Investigation regarding writing an `rrule` for using Enzyme for the ODE is ongoing, and should be the subject of next PR.

### CODE REVIEW

**Comment by VarLad** [comment]
Tests passed. If there aren't any other problems other than the ones already mentioned, this PR is ready for merge!

File: src/calculate_diff_evalue.jl [hunk 1]
```diff
@@ -1,20 +1,18 @@
...
-				
-		diff_evalue += (diff_evalue_pos - diff_evalue_neg) * map_to_circuit(gradient(_x -> phi(_x, pc, mapping), x)[1], j, mapping)
+	for j in 1:DQC.N
+		diff_evalue_pos = expect(DQC.cost, zero_state(DQC.N) => new_circuit(DQC, x, theta, j, pi/2))
+		diff_evalue_neg = expect(DQC.cost, zero_state(DQC.N) => new_circuit(DQC, x, theta, j, -pi/2))
+		diff_evalue += (diff_evalue_pos - diff_evalue_neg) * map_to_circuit(gradient(_x -> phi(_x, DQC.afm), x)[1], j, DQC.afm)
 	end
 	diff_evalue /= 2
-	return diff_evalue
+	return real(diff_evalue)
```
Comment (ChrisRackauckas) [inline src/calculate_diff_evalue.jl:17]: why real here? Is the imaginary part

File: src/phi.jl [hunk 2]
```diff
@@ -1,8 +1,8 @@
-function phi(x, ::Real, ::Product)
+function phi(x, ::Product)
 	return asin(x)
 end
 
-function phi(x, b::Real, ::Union{ChebyshevSparse, ChebyshevTower})
-	return b*acos(x)
+function phi(x, map::Union{ChebyshevSparse, ChebyshevTower})
+	return map.pc*acos(x)
```
Comment (ChrisRackauckas) [inline src/phi.jl:6]: don't shadow in local contexts, it generally leads to issues over time.

File: src/calculate_diff_evalue.jl (hunk 1 above)
Comment (VarLad) [reply src/calculate_diff_evalue.jl:17]: @ChrisRackauckas Take a look at 
`diff_evalue_pos = real.(expect(cost, zero_state(N) => new_circuit(feature_map_circuit, var, x, theta, N, pc, j, π/2; mapping)))` (old) vs  `diff_evalue_pos = expect(DQC.cost, zero_state(DQC.N) => new_circuit(DQC, x, theta, j, pi/2))` (new)

Its more efficient to convert the final value than separate values in my opinion

File: test/runtests.jl [hunk 3]
```diff
@@ -2,46 +2,16 @@ using QuantumNLDiffEq
...
-	@testset "Check if FD works" begin
-		t = 6
-		evalue(x) = expect(cost, zero_state(N)=>QuantumNLDiffEq.new_circuit(quantum_feature_map_circuit, var, x, params, N, 2; mapping=QuantumNLDiffEq.ChebyshevSparse()))
-
-		x_fd = M[t]
-		dx = 0.0001
-		x_fd += dx
-		loss_grad_fd = (evalue(x_fd) - evalue(M[t]))/dx
-
-		@test QuantumNLDiffEq.calculate_diff_evalue(quantum_feature_map_circuit, cost, var, N, params, M[t], 2; mapping=QuantumNLDiffEq.ChebyshevSparse()) ≃ loss_grad_fd
```
Comment (ChrisRackauckas) [inline test/runtests.jl:43]: why are the old tests deleted?

Comment (VarLad) [reply test/runtests.jl:43]: The tests need more work but I didn't want to write something big with some functionality missing. So the focus of this PR was to get something functional. 
After the next PR or two, I hope to add WAY better tests

File: src/calculate_diff_evalue.jl (hunk 1 above)
Comment (VarLad) [reply src/calculate_diff_evalue.jl:17]: Okay! Makes sense!

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**